### PR TITLE
opt: Eliminate unused columns using a new UsedCols property

### PR DESCRIPTION
**Pull request by andy-kimball**
This PR is broken up into multiple commits for easier review. It reorganizes and improves the FilterUsedCols rules in order to avoid creating new Project operators when they're not known to be needed.

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Thanks for the offline discussion. This is making a lot more sense now.

My biggest difficulty in digesting this is that there doesn't seem to be a principled way of setting this new property. It definitely can't be determined logically like most other properties; the implementation seems arbitrary at times (e.g. we pass through used columns in some cases but not in others). This is fine as long as we define the properly clearly (that it's just a hint, and that we set it so that things work well in practice), but it would help if we had some more context/explanation in the places where we set the property.

---

Review status: 0 of 36 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/memo/logical_props.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25277#-LBbS0o28vt0TYvvdtBh:-LBbS0o28vt0TYvvdtBi:b-30ip36) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8df75d70c37705fdd136b432753ca0d5000910/pkg/sql/opt/memo/logical_props.go#L95)):*
> ```Go
> 	OuterCols opt.ColSet
> 
> 	// UsedCols is the subset of output columns that are being referenced
> ```

If we invert it, I'd describe it like this:

PruneColsHint or PruneCandidateCols

This set is a subset of the OutputCols; it is used as a hint for pushing down
projections the tree. We try to push down a Project for columns in this set
which are not used by a higher-level operator.

From a correctness standpoint, any value (that is subset of OutputCols) is
acceptable. We set this value in a best-effort manner, in a way that interacts
best with our rules. In particular, this set should not contain the columns that
the root operator of the expression needs to determine its results (even if it
doesn't need to produce values for its output columns). For example, the
variables used by the condition of a Select are not prune candidates.

---

*[pkg/sql/opt/memo/logical_props_builder.go, line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25277#-LBaIoO3EPqvHjXMXD22:-LBbSrbf04SF8W6DPUYt:b-cd45wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8df75d70c37705fdd136b432753ca0d5000910/pkg/sql/opt/memo/logical_props_builder.go#L179)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't quite understand this. The definition is: "UsedCols is the subset of output columns that are being referenced somewhere in the expression's subtree in a way that precludes pruning". If the project is passing through columns, don't they need to be in UsedCols?
</blockquote></details>

I would rephrase this and explain that we want to push down project operations to try to remove columns that are projected but not used above the level of the Project.

---

*[pkg/sql/opt/memo/logical_props_builder.go, line 252 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25277#-LBbRE8I7Zg0icOhv_z6:-LBbRE8I7Zg0icOhv_z7:b-sko555) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b8df75d70c37705fdd136b432753ca0d5000910/pkg/sql/opt/memo/logical_props_builder.go#L252)):*
> ```Go
> 	// UsedCols are inherited from both the left and the right side. Any
> 	// additional columns used by right side (i.e. in Apply case) or the join
> 	// filter need to be counted as well.
> ```

Please add more context explaining why we are passing through used cols from the left and right side. It's not obvious, especially since other ops don't pass the columns through

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25277#-:-LBbSDTg1Bg2WEkfi2Zm:b12uhbt)*

**Comment by andy-kimball** [comment]
I'm closing this PR, since I've substantially rewritten it. I'll repost as a different PR.
